syntax = "proto3";

import "messages/basic.proto";

message Correction {
    Block start_block = 1;
    string data = 2;
}

message Corrections {
    repeated Correction corrections = 1;
}

message BlockWithSignature {
    Block block = 1;
    string strong_signature = 2;
}

message PartialMatch {
    File matched_file = 1;
    oneof optional_signature_requests {
        Blocks signature_requests = 2; 
    }
    oneof optional_corrections {
        Corrections corrections = 3;
    }
}


message SyncRequest {
    File file = 1;
    repeated uint32 weak_signatures = 2;
}

message SyncResponse {
    File requested_file = 1;
    oneof response {
        PartialMatch partial_match = 2;
        Blocks correction_request = 3;
        bool requsting_file = 4;
    }
}

message SignatureAddendum {
    repeated BlockWithSignature blocks_with_signature = 1;
}
